csrc/device_lower/analysis/tma.cpp

```diff
@@ -48,441 +50,735 @@ int64_t getCpAsyncBulkTensorSwizzleSize(TensorView* smem_tv) {
   return 1;
 }
 
-// TODO: We should use utilities in val_graph_visitor.h so that we don't have
-// to manually filter out cyclic expr groups
-ExprGroups acyclicExprGroups(const ValGraph& id_graph, const ExprGroups& egs) {
-  ExprGroups result;
-  for (const auto& eg : egs) {
-    auto inputs = id_graph.inputGroups(eg);
-    auto outputs = id_graph.outputGroups(eg);
-    bool cyclic = false;
-    for (const auto& i : inputs) {
-      for (const auto& o : outputs) {
-        if (i == o) {
-          cyclic = true;
-          goto break_two_loops;
-        }
-      }
-    }
-  break_two_loops:
-    if (!cyclic) {
-      result.pushBack(eg);
-    }
+// Infer roles (bulk, non-bulk, partitioned, box, tile, and stride) of ValGroups
+// by traversing along the traversal graph of the tensor indexer from the
+// consumer's loop domain to the gmem tensor's allocation domain. The end result
```

Comment:
Just to confirm, when you refer to "the consumer" in this file, are you referring to the shmem tensor that's the output of the TMA op?

Reply:
It refers to the smem tv in TMA load, and the gmem tv in TMA store, this pass handle both load and store the same way.